satellite/metabase/rangedloop/service.go

```diff
@@ -74,81 +86,116 @@ func (service *Service) Run(ctx context.Context) (err error) {
-
-	for _, obs := range service.observers {
-		err := obs.Start(ctx, startTime)
-		if err != nil {
-			return err
-		}
-
-		observerStates = append(observerStates, observerState{
-			observer: obs,
-		})
+	observerStates, err := startObservers(ctx, service.observers)
+	if err != nil {
+		return nil, err
 	}
 
 	rangeProviders, err := service.provider.CreateRanges(service.config.Parallelism, service.config.BatchSize)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	group := errs2.Group{}
-	for rangeIndex, rangeProvider := range rangeProviders {
-		rangeObservers := []Partial{}
+	for _, rangeProvider := range rangeProviders {
+		rangeObservers := []*rangeObserverState{}
 		for i, observerState := range observerStates {
 			rangeObserver, err := observerState.observer.Fork(ctx)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			rangeObservers = append(rangeObservers, rangeObserver)
-			observerStates[i].rangeObservers = append(observerStates[i].rangeObservers, rangeObserver)
-		}
-
-		// Create closure to capture loop variables.
-		createClosure := func(ctx context.Context, rangeIndex int, rangeProvider SegmentProvider, rangeObservers []Partial) func() error {
-			return func() (err error) {
-				defer mon.Task()(&ctx)(&err)
-
-				return rangeProvider.Iterate(ctx, func(segments []segmentloop.Segment) error {
-					for _, rangeObserver := range rangeObservers {
-						err := rangeObserver.Process(ctx, segments)
-						if err != nil {
-							return err
-						}
-					}
-					return nil
-				})
+			rangeState := &rangeObserverState{
+				rangeObserver: rangeObserver,
+				duration:      time.Duration(0),
 			}
+			rangeObservers = append(rangeObservers, rangeState)
+			observerStates[i].rangeObservers = append(observerStates[i].rangeObservers, rangeState)
 		}
 
-		group.Go(createClosure(ctx, rangeIndex, rangeProvider, rangeObservers))
+		// Create closure to capture loop variables.
+		group.Go(createGoroutineClosure(ctx, rangeProvider, rangeObservers))
 	}
 
 	// Improvement: stop all ranges when one has an error.
 	errList := group.Wait()
 	if errList != nil {
-		return errs.Combine(errList...)
+		return nil, errs.Combine(errList...)
 	}
 
-	// Segment loop has ended.
-	// This is the reduce step.
-	for _, state := range observerStates {
-		for _, rangeObserver := range state.rangeObservers {
-			err := state.observer.Join(ctx, rangeObserver)
-			if err != nil {
-				return err
+	return finishObservers(ctx, observerStates)
+}
+
+func createGoroutineClosure(ctx context.Context, rangeProvider SegmentProvider, states []*rangeObserverState) func() error {
+	return func() (err error) {
+		defer mon.Task()(&ctx)(&err)
+
+		return rangeProvider.Iterate(ctx, func(segments []segmentloop.Segment) error {
+			for i, state := range states {
+				start := time.Now()
+				err := state.rangeObserver.Process(ctx, segments)
+				if err != nil {
+					return err
+				}
+				states[i].duration += time.Since(start)
```

Comment:
nitpick: since state is a *rangeObserverState, you can mutate directly:
```suggestion
				state.duration += time.Since(start)
```

Reply:
Done